From b204eda2ea08c1bd15ddcf541c9b26821e5ab2cf Mon Sep 17 00:00:00 2001
Date: Thu, 14 Sep 2023 14:53:50 +0800
Subject: [PATCH] ascend: add emmc patch for helper

Offering:ascend
Reference:NA
Type:requirement
DTS/AR:AR000231NC
reason:helper support emmc ras
---
 drivers/mmc/host/dw_mmc.c | 35 +++++++++++++++++++++++++++++++++++
 drivers/mmc/host/dw_mmc.h |  1 +
 2 files changed, 36 insertions(+)

diff --git a/drivers/mmc/host/dw_mmc.c b/drivers/mmc/host/dw_mmc.c
index 22c454c7a..8fbca613c 100644
--- a/drivers/mmc/host/dw_mmc.c
+++ b/drivers/mmc/host/dw_mmc.c
@@ -50,6 +50,8 @@
 				 SDMMC_INT_RESP_ERR | SDMMC_INT_HLE)
 #define DW_MCI_ERROR_FLAGS	(DW_MCI_DATA_ERROR_FLAGS | \
 				 DW_MCI_CMD_ERROR_FLAGS)
+#define DW_MCI_RAS_ERROR_FLAGS	(SDMMC_INT_RESP_ERR | SDMMC_INT_RCRC | SDMMC_INT_DCRC | SDMMC_INT_DRTO | \
+				 SDMMC_INT_HTO | SDMMC_INT_FRUN | SDMMC_INT_HLE | SDMMC_INT_SBE | SDMMC_INT_EBE)
 #define DW_MCI_SEND_STATUS	1
 #define DW_MCI_RECV_STATUS	2
 #define DW_MCI_DMA_THRESHOLD	16
@@ -2637,6 +2639,38 @@ static void dw_mci_handle_cd(struct dw_mci *host)
 		msecs_to_jiffies(host->pdata->detect_delay_ms));
 }
 
+DEFINE_SPINLOCK(emmcras_irq_lock);
+static int emmc_ras_err_flags = 0;
+int dw_mci_get_emmc_ras_reg_data_flags(void)
+{
+	int ras_err = 0;
+	unsigned long flags;
+
+	spin_lock_irqsave(&emmcras_irq_lock, flags);
+	ras_err = emmc_ras_err_flags;
+	emmc_ras_err_flags = 0;
+	spin_unlock_irqrestore(&emmcras_irq_lock, flags);
+
+	return ras_err;
+}
+EXPORT_SYMBOL(dw_mci_get_emmc_ras_reg_data_flags);
+
+static void dw_mci_set_emmc_ras_reg_data_flags(struct dw_mci *host, u32 pending)
+{
+	unsigned long flags;
+
+	spin_lock_irqsave(&emmcras_irq_lock, flags);
+	emmc_ras_err_flags |= (pending & DW_MCI_RAS_ERROR_FLAGS);
+
+	/* Check volt switch first, since it can look like an error */
+	if ((host->state == STATE_SENDING_CMD11) && (pending & SDMMC_INT_VOLT_SWITCH)) {
+			emmc_ras_err_flags &= ~SDMMC_INT_VOLT_SWITCH;
+	}
+	spin_unlock_irqrestore(&emmcras_irq_lock, flags);
+
+	return;
+}
+
 static irqreturn_t dw_mci_interrupt(int irq, void *dev_id)
 {
 	struct dw_mci *host = dev_id;
@@ -2645,6 +2679,7 @@ static irqreturn_t dw_mci_interrupt(int irq, void *dev_id)
 	unsigned long irqflags;
 
 	pending = mci_readl(host, MINTSTS); /* read-only mask reg */
+	dw_mci_set_emmc_ras_reg_data_flags(host, pending);
 
 	if (pending) {
 		/* Check volt switch first, since it can look like an error */
diff --git a/drivers/mmc/host/dw_mmc.h b/drivers/mmc/host/dw_mmc.h
index 46e9f8ec5..77f43889f 100644
--- a/drivers/mmc/host/dw_mmc.h
+++ b/drivers/mmc/host/dw_mmc.h
@@ -500,6 +500,7 @@ struct dw_mci_board {
 
 extern int dw_mci_probe(struct dw_mci *host);
 extern void dw_mci_remove(struct dw_mci *host);
+int dw_mci_get_emmc_ras_reg_data_flags(void);
 #ifdef CONFIG_PM
 extern int dw_mci_runtime_suspend(struct device *device);
 extern int dw_mci_runtime_resume(struct device *device);
-- 
2.27.0

